src/lib.rs

```diff
@@ -321,106 +320,143 @@ impl Check {
-                let names = match &self.scope.mode {
-                    ScopeMode::DenyList(_) =>
-                        match &self.current.source {
-                            RustdocSource::Rustdoc(_) =>
-                                vec!["the-name-doesnt-matter-here".to_string()],
-                            _ => panic!("couldn't deduce crate name, specify one through the package allow list")
-                        }
-                    ScopeMode::AllowList(lst) => lst.clone(),
-                };
-                names
-                    .iter()
-                    .map(|name| {
-                        let version = None;
-                        let (current_crate, baseline_crate) = generate_versioned_crates(
-                            &mut config,
-                            &rustdoc_cmd,
-                            &*current_loader,
-                            &*baseline_loader,
-                            name,
-                            version,
-                        )?;
-
-                        let success = run_check_release(
-                            &mut config,
-                            name,
-                            current_crate,
-                            baseline_crate,
-                            self.release_type,
-                        )?;
-                        Ok(success)
-                    })
-                    .collect()
-            }
-            RustdocSource::Root(project_root) => {
-                let metadata = manifest_metadata(project_root)?;
-                let selected = self.scope.selected_packages(&metadata);
-                selected
-                    .iter()
-                    .map(|selected| {
-                        let crate_name = &selected.name;
-                        let version = &selected.version;
-
-                        // If the manifest we're using points to a workspace, then
-                        // ignore `publish = false` crates unless they are specifically selected.
-                        // If the manifest points to a specific crate, then check the crate
-                        // even if `publish = false` is set.
-                        let is_implied = matches!(self.scope.mode, ScopeMode::DenyList(..))
-                            && metadata.workspace_members.len() > 1
-                            && selected.publish == Some(vec![]);
-                        if is_implied {
-                            config.verbose(|config| {
-                                config.shell_status(
-                                    "Skipping",
-                                    format_args!("{crate_name} v{version} (current)"),
-                                )
-                            })?;
-                            Ok(true)
-                        } else {
+        // Create a report for each crate.
+        let all_outcomes: anyhow::Result<BTreeMap<String, anyhow::Result<CrateReport>>> =
```

Comment:
Two concerns here:
- I feel like this code has gotten a bit too complex and intricate. It took me probably 10-15 minutes to figure out what exactly is going on, and that's when I was already familiar with what this code was doing. We're adding significant new complexity here — do you think there's a reasonable refactor we could do to improve the readability in the process?
- The `Result<Map<String, Result<CrateReport>>>` type is quite confusing, especially because it isn't documented what becomes an inner error vs an outer error. It seems that failure to generate the rustdoc becomes an outer error? Is that right? And if so, should that be the case? I'm genuinely not sure.

Reply:
> We're adding significant new complexity here 

I agree. But trying to address [this](https://github.com/obi1kenobi/cargo-semver-checks/pull/398#discussion_r1117872436) is not easy, because you can't simply `?` everywhere when you have a BTreeMap.

> do you think there's a reasonable refactor we could do to improve the readability in the process?

I will try to think about it and change that big Result type.